File: web/src/recognition/recon.js

// Cargar AWS SDK
import AWS from 'aws-sdk';

// Cargar el módulo de FileSystem para leer archivos locales
import fs from 'fs';

// Configurar AWS Rekognition con las credenciales del archivo .env
AWS.config.update({
  accessKeyId: process.env.AWS_ACCESS_KEY_ID,
  secretAccessKey: process.env.AWS_SECRET_ACCESS_KEY,
  region: process.env.AWS_DEFAULT_REGION
});

// Crear una instancia de AWS Rekognition
const rekognition = new AWS.Rekognition();

export const detectDominantEmotionFromLocalImage = async (imagePath) => {
  try {
    // Leer la imagen local y convertirla en un buffer
    const imageBytes = fs.readFileSync(imagePath);

    // Parámetros para la API de Rekognition
    const params = {
      Image: {
        Bytes: imageBytes
      },
      Attributes: ['ALL']
    };

    // Llamar a Rekognition para detectar emociones
    const data = await rekognition.detectFaces(params).promise();
    
    let dominantEmotion = null;
    let maxConfidence = 0;

    // Buscar la emoción con la confianza más alta
    data.FaceDetails.forEach((faceDetail) => {
      faceDetail.Emotions.forEach((emotion) => {
        if (emotion.Confidence > maxConfidence) {
          maxConfidence = emotion.Confidence;
          dominantEmotion = emotion.Type;
        }
      });
    });

    return dominantEmotion ? { "status": dominantEmotion } : 'No se pudo detectar ninguna emoción';
  } catch (err) {
    console.error("Error al detectar emociones:", err);
    throw err;
  }
};

// Ejemplo de uso:
// detectDominantEmotionFromLocalImage('img/test.jpg')
//   .then(dominantEmotion => console.log("Status:", dominantEmotion))
//   .catch(err => console.error(err));

File: server/recon.py
import boto3
import os
from dotenv import load_dotenv

load_dotenv()

aws_access_key_id = os.getenv('AWS_ACCESS_KEY_ID')
aws_secret_access_key = os.getenv('AWS_SECRET_ACCESS_KEY')
aws_region = os.getenv('AWS_REGION')
# Configurar el cliente de AWS Rekognition

rekognition = boto3.client('rekognition', 
    region_name=aws_region, 
    aws_access_key_id=aws_access_key_id, 
    aws_secret_access_key=aws_secret_access_key
)

def detect_dominant_emotion(image_bytes):
    try:
        response = rekognition.detect_faces(
            Image={'Bytes': image_bytes},
            Attributes=['ALL']
        )
        face_details = response.get('FaceDetails', [])
        
        if not face_details:
            return 'No se pudo detectar ninguna emoción'

        # Encontrar la emoción más predominante
        dominant_emotion = None
        max_confidence = 0

        for face in face_details:
            for emotion in face['Emotions']:
                if emotion['Confidence'] > max_confidence:
                    max_confidence = emotion['Confidence']
                    dominant_emotion = emotion['Type']

        return dominant_emotion if dominant_emotion else 'No se pudo detectar ninguna emoción'
    except Exception as e:
        print(f'Error al detectar emociones: {e}')
        return 'Error al procesar la imagen'
